ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py

```diff
@@ -0,0 +1,157 @@
+    UnknownMetricFamily,
+)
+
+from opentelemetry.metrics import Counter, Gauge, Measure, Metric
+from opentelemetry.sdk.metrics.export import (
+    MetricRecord,
+    MetricsExporter,
+    MetricsExportResult,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class PrometheusMetricsExporter(MetricsExporter):
+    """Prometheus metric exporter for OpenTelemetry.
+
+    Args:
+        port: The Prometheus port to be used.
+        address: The Prometheus address to be used.
+        prefix: single-word application prefix relevant to the domain the metric belongs to.
+    """
+
+    def __init__(self, port: int = 8000, address: str = "", prefix: str = ""):
+        self._port = port
+        self._address = address
+        self._collector = CustomCollector(prefix)
+
+        start_http_server(port=self._port, addr=str(self._address))
+        REGISTRY.register(self._collector)
+
+    def export(
+        self, metric_records: Sequence[MetricRecord]
+    ) -> MetricsExportResult:
+        self._collector.add_metrics_data(metric_records)
+        return MetricsExportResult.SUCCESS
+
+    def shutdown(self) -> None:
+        REGISTRY.unregister(self._collector)
+
+
+class CustomCollector:
+    """ CustomCollector represents the Prometheus Collector object
+        https://github.com/prometheus/client_python#custom-collectors
+    """
+
+    def __init__(self, prefix: str = ""):
+        self._prefix = prefix
+        self._metrics_to_export = []
+
+    def add_metrics_data(self, metric_records: Sequence[MetricRecord]):
+        self._metrics_to_export.append(metric_records)
+
+    def collect(self):
+        """Collect fetches the metrics from OpenTelemetry
+        and delivers them as Prometheus Metrics.
+        Collect is invoked every time a prometheus.Gatherer is run
+        for example when the HTTP endpoint is invoked by Prometheus.
+        """
+
+        for metric_batch in self._metrics_to_export:
+            for metric_record in metric_batch:
+                prometheus_metric = self._translate_to_prometheus(
+                    metric_record
+                )
+                if prometheus_metric:
+                    yield prometheus_metric
+            self._metrics_to_export.remove(metric_batch)
+
+    def _translate_to_prometheus(self, metric_record: MetricRecord):
+        prometheus_metric = None
+        label_values = []
+        label_keys = []
+        for label in metric_record.label_set.labels:
+            for index, label_tuple_value in enumerate(label):
+                # Odd number
+                if index & 1 == 1:
+                    label_values.append(label_tuple_value)
+
+        for label_key in metric_record.metric.label_keys:
+            label_keys.append(sanitize(label_key))
+        metric_name = ""
+        if self._prefix != "":
+            metric_name = self._prefix + "_"
+        metric_name += sanitize(metric_record.metric.name)
+
+        if isinstance(metric_record.metric, Counter):
```

Comment:
This looks pretty cool, but I'm not sure I understand how this works exactly can you explain this a little bit more?

With this code I'm currently getting a KeyError: <class 'opentelemetry.sdk.metrics.Counter'>

Reply:
Ok, a solution is:

```python
        if isinstance(metric_record.metric, (Counter, Gauge, Measure)):
            prometheus_metric = {
                Counter: CounterMetricFamily,
                Gauge: GaugeMetricFamily,
                Measure: UnknownMetricFamily
            }[metric_record.metric.__class__.__bases__[1]](
                name=metric_name,
                documentation=metric_record.metric.description,
                labels=label_keys,
            )
            prometheus_metric.add_metric(
                labels=label_values, value=metric_record.aggregator.check_point
            )

        else:
            logger.warning(
                "Unsupported metric type. %s", type(metric_record.metric)
            )
```

This was failing because the class of `metric_record.metric` was a child of `Gauge` instead of being `Gauge` itself. It is just a suggestion, you may not find it worthy to dig that deep into the attributes of an object to avoid some repeated code.

Reply:
Let me know if you have strong opinions about this one, I agree we will have less repeated code but we are adding more complicated one.

Reply:
Not really, feel free to use this suggestion or not as you like :wink: